### ISSUE TITLE
BUG: Slow down in function evaluations within scipy.optimize.minimize after 1.15.0

### ISSUE THREAD

**Issue by saitcakmak** [scipy/scipy#22438]
### Describe your issue.

We have been seeing significant slowdowns (~4x increase in runtime) in our (pytorch/botorch) GitHub CI after the release of scipy 1.15.0. On closer inspection, I've isolated the slowdowns to the function evaluations within `scipy.optimize.minimize` calls, which in some cases were 10x slower than before. I'll share a simplified repro using only scipy & torch below, that reproduces a slowdown of a smaller magnitude.

Things to note:
- We've only been able to reproduce the issue reliably in GitHub actions (e.g, using 4-core-ubuntu runner) & in google colab to some extent. The issue does not reproduce on a CPU with high core count (such as a macbook w/ M1 pro, or a server running CentOS 9 with 32 core Intel Xeon Skylake processor). 
- The slowdown only happens when using L-BFGS-B optimizer (and only with gradient observations), which has been ported over to C in 1.15.0. We also use SLSQP, which does not show any slowdown.
- The slowdown goes away (in the sense that measurement inside & outside of scipy.minimize are equivalent) if I set `OMP_NUM_THREADS=1, OPENBLAS_NUM_THREADS=1, MKL_NUM_THREADS=1, VECLIB_MAXIMUM_THREADS=1, NUMEXPR_NUM_THREADS=1` (I didn't verify whether all of them are necessary). This was tested in GitHub actions, did not attempt with colab.

I've ran the codeblock below with `!pip install scipy==1.14.1 torch` and `!pip install scipy==1.15.1 torch` on google colab and observed following runtimes (taken from one run, reproduced similarly in other runs). The runtimes are taken from cProfile output, the `cumtime` column of `f_and_grad` row.

With scipy 1.15.1 (~47% slower):
- 13.593 seconds for 102 evaluations of `f_and_grad` within `minimize` call
- 9.061 seconds for 100 reference evaluations of `f_and_grad` in a loop

With scipy 1.14.1 (roughly equivalent):
- 8.923 seconds for 101 evaluations of `f_and_grad` within `minimize` call
- 8.911 seconds for 100 reference evaluations of `f_and_grad` in a loop

### Reproducing Code Example

```python
# Setup the function & optimization parameters

import torch
import numpy as np
from scipy.optimize import minimize, Bounds
import cProfile, io, pstats
from pstats import SortKey

sortby = SortKey.CUMULATIVE
torch.manual_seed(0)
np.random.seed(0)
torch.set_default_dtype(torch.double)

bounds = Bounds(
    lb=np.zeros(128),
    ub=np.ones(128),
    keep_feasible=True,
)
x0 = np.random.random(128)
rand_tensor = torch.rand(128, 128, 128)

def f_and_grad(x):
    X_tensor = torch.from_numpy(x).contiguous().requires_grad_(True)
    diff = rand_tensor - X_tensor
    loss = (diff @ diff).sum()
    grad = torch.autograd.grad(loss, X_tensor)[0].reshape(-1).numpy(force=True)
    return loss.item(), grad



# Discard this one. The first call tends to be slower than subsequent calls.
with cProfile.Profile() as pr:
    res = minimize(
        fun=f_and_grad,
        x0=x0,
        method="L-BFGS-B",
        jac=True,
        bounds=bounds,
        options={"maxfun": 20, "ftol": 0},
    )


# Profile with minimize -- runtime of this cell differs significantly between 1.15.1 and 1.14.1
with cProfile.Profile() as pr:
    res = minimize(
        fun=f_and_grad,
        x0=x0,
        method="L-BFGS-B",
        jac=True,
        bounds=bounds,
        options={"maxfun": 100, "ftol": 0},
    )
ps = pstats.Stats(pr).sort_stats(sortby)
ps.print_stats(.1)


# Reference evaluations
X = np.random.random(128)

with cProfile.Profile() as pr:
    for _ in range(100):
        f_and_grad(X)

ps = pstats.Stats(pr).sort_stats(sortby)
ps.print_stats(.1)
```

### Error message

```shell
Example output on 1.15.1:

# Profile with minimize

         11385 function calls in 13.879 seconds

   Ordered by: cumulative time
   List reduced from 137 to 14 due to restriction <0.1>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   13.879   13.879 /usr/local/lib/python3.11/dist-packages/scipy/optimize/_minimize.py:53(minimize)
        1    0.007    0.007   13.877   13.877 /usr/local/lib/python3.11/dist-packages/scipy/optimize/_lbfgsb_py.py:290(_minimize_lbfgsb)
      204    0.144    0.001   13.744    0.067 /usr/local/lib/python3.11/dist-packages/scipy/optimize/_optimize.py:71(_compute_if_needed)
      103    0.001    0.000   13.742    0.133 /usr/local/lib/python3.11/dist-packages/scipy/optimize/_differentiable_functions.py:293(_update_fun)
      102    0.002    0.000   13.741    0.135 /usr/local/lib/python3.11/dist-packages/scipy/optimize/_differentiable_functions.py:16(wrapped)
      102    0.001    0.000   13.737    0.135 /usr/local/lib/python3.11/dist-packages/scipy/optimize/_optimize.py:78(__call__)
      102    3.654    0.036   13.593    0.133 <timed exec>:5(f_and_grad)
      102    0.002    0.000   13.574    0.133 /usr/local/lib/python3.11/dist-packages/scipy/optimize/_differentiable_functions.py:341(fun_and_grad)
      102    0.006    0.000    9.699    0.095 /usr/local/lib/python3.11/dist-packages/torch/autograd/__init__.py:358(grad)
      102    0.002    0.000    9.676    0.095 /usr/local/lib/python3.11/dist-packages/torch/autograd/graph.py:816(_engine_run_backward)
      102    9.674    0.095    9.674    0.095 {method 'run_backward' of 'torch._C._EngineBase' objects}
      102    0.227    0.002    0.227    0.002 {method 'sum' of 'torch._C.TensorBase' objects}
        1    0.000    0.000    0.191    0.191 /usr/local/lib/python3.11/dist-packages/scipy/optimize/_optimize.py:203(_prepare_scalar_function)
        1    0.000    0.000    0.191    0.191 /usr/local/lib/python3.11/dist-packages/scipy/optimize/_differentiable_functions.py:166(__init__)


CPU times: user 17.5 s, sys: 2.26 s, total: 19.8 s


# Reference evaluations
         4602 function calls in 9.061 seconds

   Ordered by: cumulative time
   List reduced from 29 to 3 due to restriction <0.1>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      100    2.408    0.024    9.061    0.091 <timed exec>:5(f_and_grad)
      100    0.005    0.000    6.459    0.065 /usr/local/lib/python3.11/dist-packages/torch/autograd/__init__.py:358(grad)
      100    0.002    0.000    6.440    0.064 /usr/local/lib/python3.11/dist-packages/torch/autograd/graph.py:816(_engine_run_backward)


CPU times: user 8.12 s, sys: 994 ms, total: 9.11 s
```

### SciPy/NumPy/Python version and system information

```shell
On colab with scipy 1.15.1:
1.15.1 1.26.4 sys.version_info(major=3, minor=11, micro=11, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /opt/_internal/cpython-3.11.10/lib/python3.11/site-packages/scipy_openblas32/include
    lib directory: /opt/_internal/cpython-3.11.10/lib/python3.11/site-packages/scipy_openblas32/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.28 DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=64
    pc file directory: /project
    version: 0.3.28
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /opt/_internal/cpython-3.11.10/lib/python3.11/site-packages/scipy_openblas32/include
    lib directory: /opt/_internal/cpython-3.11.10/lib/python3.11/site-packages/scipy_openblas32/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.28 DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=64
    pc file directory: /project
    version: 0.3.28
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.13.6
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.11
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/pip-build-env-63dv7byy/overlay/lib/python3.11/site-packages/pythran
    version: 0.17.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/python/cp311-cp311/bin/python
  version: '3.11'
```

**Comment by ilayn** [comment]
If the results are the same (and looks like the eval counts are comparable) then the only detail that probably makes the difference is the OpenBLAS (such as `OPENBLAS_NUM_THREADS=1` you found out). Otherwise the code is pretty much the same and I did not observe any speed difference between the F77 and C code performance back when it was ported.

I think the easiest is to try it in a conda env by switching to MKL to see whether it makes a difference.

**Comment by saitcakmak** [comment]
Thanks @ilayn! If I use `conda install pytorch scipy==1.15.1 -c conda-forge` (verified that MKL is installed) for the setup, the slowdown goes away. Any suggestions on how to prevent the slowdown when conda / MKL is not available? 

I've tested this on an AWS t2xlarge instance (4vCPU, 16G ram). On an `uv`-based environment, the runtime difference was 15.497 secs (within minimize) vs 7.397 secs (without minimize). The runtimes are pretty much equal with scipy 1.14.1 or when using conda w/ MKL as suggested.

**Comment by ilayn** [comment]
So if I am understanding correctly, switching to MKL made the issue disappear (correct me if I am wrong). 

The main difference with 1.14.1 and 1.15.x is that we ported the solver to C and updated the algorithm to make proper LAPACK calls instead of the homebaked F77 code[^1] . Hence, now that the new code is making LAPACK calls, you are being hit by LAPACK/OpenBLAS issues that you were being hit before.

I am not well-versed in the environments you mentioned above, but I would start with making sure that OpenBLAS that is being used is single threaded. The noticeable slowdowns typically come from unmatched thread settings. we do get occasional slowdown reports at every version release which is typically solved by exporting an environment var

```powershell
$env:OPENBLAS_NUM_THREADS=1
```

There are a lot of similar fun issues, in case you are interested, such as #22105 or #21479. Hopefully other folks can chime in about the specifics for the architectures you have been using.


[^1]: Not that it was wrong but outdated by a few decades, not handling overflows etc.

**Comment by saitcakmak** [comment]
> So if I am understanding correctly, switching to MKL made the issue disappear (correct me if I am wrong).

Yes, that's correct.

> I would start with making sure that OpenBLAS that is being used is single threaded.

Setting `OPENBLAS_NUM_THREADS=1` does resolve the slowdown in this particular example. 

Thanks for the help so far! We could set the env var for our CI, but it doesn't seem like a solution that we could easily extend to our users out of the box. I wouldn't want to set `OPENBLAS_NUM_THREADS=1` globally on import and risk slowing down user code elsewhere. I'd appreciate any input on how to fix this more generally without globally modifying the environment variables.

**Comment by lucascolley** [comment]
To understand what's going on here, see https://pythonspeed.com/articles/concurrency-control/.

**Comment by lucascolley** [comment]
> We could set the env var for our CI, but it doesn't seem like a solution that we could easily extend to our users out of the box. I wouldn't want to set OPENBLAS_NUM_THREADS=1 globally on import and risk slowing down user code elsewhere. I'd appreciate any input on how to fix this more generally without globally modifying the environment variables.

https://github.com/joblib/threadpoolctl is probably what you are looking for.

**Comment by saitcakmak** [comment]
> https://github.com/joblib/threadpoolctl is probably what you are looking for.

Thanks for the suggestion! This does seem to work and doesn't affect PyTorch AFAICT.

I'll close the issue since we've found a solution. For anyone looking at this in the future, I've used
```
from threadpoolctl import threadpool_limits

with threadpool_limits(limits=1, user_api='blas'):
    res = minimize(...)
```
PyTorch relies on OpenMP & MKL rather than OpenBLAS, so it doesn't seem to be affected by the context manager here.

**Comment by nabenabe0928** [comment]
```python
import math
import numpy as np
import scipy
import torch
import time


n_points = 128
seed = 5
torch.manual_seed(seed)
np.random.seed(seed)
rand_tensor = torch.rand(n_points, n_points, n_points)
counter = 0


def func(x):
    global counter
    counter += 1
    start = time.time()
    X_tensor = torch.from_numpy(x).requires_grad_(True)
    diff = rand_tensor - X_tensor
    loss = (diff @ diff).sum()
    grad = torch.autograd.grad(loss, X_tensor)[0].reshape(-1).numpy(force=True)
    print(f"Func call {counter:0>2}:", time.time() - start)
    return loss.item(), grad


x0 = np.random.random(n_points)
start = time.time()
res = scipy.optimize.minimize(
    func,
    x0.copy(),
    jac=True,
    method="l-bfgs-b",
    options={"gtol": 1e-2},
)
print(time.time() - start, scipy.__version__)

```

I used the code above to reproduce the slowdown.
I am observing the slowdown only from the 38-th to 51-th calls and this phenomenon reproduces every time I try, which seems very deterministic and does not explain the cause by multi-threading.

I suspect there are some other reasons why it slows down.

The shared environment is as follows:

- OS: Ubuntu 20.04
- Python Version: 3.12.11
- Python Package Dependencies:

```
filelock==3.13.1
fsspec==2024.6.1
Jinja2==3.1.4
MarkupSafe==2.1.5
mpmath==1.3.0
networkx==3.3
numpy==2.2.6
setuptools==70.2.0
sympy==1.13.3
torch==2.7.1+cpu
typing_extensions==4.12.2
```

The only difference in the benchmarking was SciPy, which I used `scipy==1.14.1` and `scipy==1.15.3`.

> [!NOTE]
> Although the slowdown became milder with the suggestion above, which is to limit the thread count by exporting `OPENBLAS_NUM_THREADS=1`, I think there are some additional causes of the slowdown.

> [!IMPORTANT]
> In my environment, `with threadpool_limits(limits=1, user_api="blas")` did not remove the slowdown with `threadpoolctl==3.6.0`.

**Comment by nabenabe0928** [comment]
@ilayn @lucascolley 
Sorry, I forgot to mention Scipy developers in the comment above 🙇

**Comment by nabenabe0928** [comment]
Another solution (if you are not using any operations that use multi-threading in combination with OpenBLAS):

```python
from contextlib import contextmanager
import os


@contextmanager
def single_blas_thread() -> None:
    old_val = os.environ.get("OPENBLAS_NUM_THREADS")
    os.environ["OPENBLAS_NUM_THREADS"] = "1"
    try:
        yield
    finally:
        if old_val is None:
            os.environ.pop("OPENBLAS_NUM_THREADS", None)
        else:
            os.environ["OPENBLAS_NUM_THREADS"] = old_val


with single_blas_thread():
    res = scipy.optimize.minimize(...)
```

**Comment by ilayn** [comment]
Hi @nabenabe0928 do you mind opening a new issue for this? The slowdowns are probably not component related but internally doing more iterations. Something in Ubuntu based systems (so far the only detail we could identify as the common point) is causing this issue as reported in #23161

**Comment by nabenabe0928** [comment]
@ilayn 

Hey, thank you for the suggestion, I opened another issue as asked:
- https://github.com/scipy/scipy/issues/23191
